sympy/functions/elementary/piecewise.py

```diff
@@ -1138,25 +1017,24 @@ def _clip(A, B, k):
     # pair(A, B) will give keys of -1, 0 and 1
     # we rekey A intervals to k and change all others
     # to -1
+    assert k != -1
     a, b = B
     c, d = A
     B = a, b = Min(a, b), b
-    A = c, d = Min(Max(c, a), b), Min(Max(d, a), b)
-    p = _pair(A, B, abi=True)
-    # remove infinities that weren't there at the start
+    c, d = Min(Max(c, a), b), Min(Max(d, a), b)
+    A = c, d = Min(c, d), d
     oo = S.Infinity
-    if -oo not in (a, c) and p[0][0] == -oo:
-        p.pop(0)
-    if oo not in (b, d) and p[-1][1] == oo:
-        p.pop()
-    # rekey key 0 to k
-    assert k != -1
-    for i in range(len(p)):
-        if p[i][-1] == 0:
-            p[i] = p[i][:2] + (k,)
-        else:
-            # rekey all other keys to -1
-            p[i] = p[i][:2] + (-1,)
+    p = []
+    if a is -oo and c is not -oo:
+        p.append((-oo, c, -1))
+    if c != d:
+        p.append((c, d, k))
+    if d != a:
+        p.append((d, a, -1))
```

Comment:
Do we now have `a <= d <= b`?

Reply:
`a <= b` b/c `a, b = Min(a, b), b` at lone 1023; and at 1024, `d = Min(Max(d, a), b)` -- that's not enough for me to see it (though I can slowly reason it out). I test that as:

```
>>> for i in permutations((1,2,3)):
...     a, b, d = i
...     a = min(a, b)
...     d =  Min(Max(d, a), b)
...     assert a <= d and d <= b
```

Reply:
When I can use an interactive session I will track down the failures and see what assumptions I've violated. I put the title back to [wip].